core-bundle/src/EventListener/DataContainer/PageUrlListener.php

```diff
@@ -0,0 +1,397 @@
+     */
+    public function validateUrlPrefix(string $value, DataContainer $dc): string
+    {
+        if ('root' !== $dc->activeRecord->type || $dc->activeRecord->urlPrefix === $value) {
+            return $value;
+        }
+
+        // First check if another root page uses the same url prefix and domain
+        $count = $this->connection
+            ->executeQuery(
+                'SELECT COUNT(*) FROM tl_page WHERE urlPrefix=:urlPrefix AND dns=:dns AND id!=:rootId',
+                [
+                    'urlPrefix' => $value,
+                    'dns' => $dc->activeRecord->dns,
+                    'rootId' => $dc->id,
+                ]
+            )
+            ->fetchColumn()
+        ;
+
+        if ($count > 0) {
+            throw new \RuntimeException($this->translator->trans('ERR.urlPrefixExists', [$value], 'contao_default'));
+        }
+
+        /** @var PageModel $pageAdapter */
+        $pageAdapter = $this->framework->getAdapter(PageModel::class);
+        $rootPage = $pageAdapter->findByPk($dc->id);
+
+        if (null === $rootPage) {
+            return $value;
+        }
+
+        try {
+            $this->recursiveValidatePages((int) $rootPage->id, $rootPage);
+        } catch (DuplicateAliasException $exception) {
+            throw new \RuntimeException($this->translator->trans('ERR.pageUrlPrefix', [$exception->getUrl()], 'contao_default'), $exception->getCode(), $exception);
+        }
+
+        return $value;
+    }
+
+    /**
+     * @Callback(table="tl_page", target="fields.urlSuffix.save")
+     */
+    public function validateUrlSuffix($value, DataContainer $dc)
+    {
+        if ('root' !== $dc->activeRecord->type || $dc->activeRecord->urlSuffix === $value) {
+            return $value;
+        }
+
+        /** @var PageModel $pageAdapter */
+        $pageAdapter = $this->framework->getAdapter(PageModel::class);
+        $rootPage = $pageAdapter->findByPk($dc->id);
+
+        if (null === $rootPage) {
+            return $value;
+        }
+
+        try {
+            $this->recursiveValidatePages((int) $rootPage->id, $rootPage);
+        } catch (DuplicateAliasException $exception) {
+            throw new \RuntimeException($this->translator->trans('ERR.pageUrlSuffix', [$exception->getUrl()], 'contao_default'), 0, $exception);
+        }
+
+        return $value;
+    }
+
+    public function reset(): void
+    {
+        $this->prefixes = null;
+        $this->suffixes = null;
+    }
+
+    private function purgeSearchIndex(int $pageId): void
+    {
+        $urls = $this->connection
+            ->executeQuery(
+                'SELECT url FROM tl_search WHERE pid=:pageId',
+                ['pageId' => $pageId]
+            )
+            ->fetchAll(FetchMode::COLUMN)
+        ;
+
+        foreach ($urls as $url) {
+            $this->searchIndexer->delete(new Document(new Uri($url), 200));
+        }
+    }
+
+    /**
+     * @throws DuplicateAliasException
```

Comment:
Not thrown, therefore please remove.

Reply:
it is, because the method calls `$this->aliasExists()`

Reply:
Well, then the `aliasExists()` method throws the exception but no the `recursiveValidatePages()` method. We are not following the "every method must annotate all exceptions thrown in every child method" path.

In fact, we are not requiring `@throws` annotations at all.